usersession/agent/rest_api.go

```diff
@@ -197,3 +209,112 @@ func postServiceControl(c *Command, r *http.Request) Response {
+		TimeRemaining       time.Duration `json:"time-remaining,omitempty"`
+		BusyAppName         string        `json:"busy-app-name,omitempty"`
+		BusyAppDesktopEntry string        `json:"busy-app-desktop-entry,omitempty"`
+	}
+	var refreshInfo pendingSnapRefreshInfo
+	if err := decoder.Decode(&refreshInfo); err != nil {
+		return BadRequest("cannot decode request body into pending snap refresh info: %v", err)
+	}
+
+	conn, err := dbusutil.SessionBus()
+	if err != nil {
+		return SyncResponse(&resp{
+			Type:   ResponseTypeError,
+			Status: 500,
+			Result: &errorResult{
+				Message: fmt.Sprintf("cannot connect to the session bus: %v", err),
+			},
+		})
+	}
+	defer conn.Close()
+
+	// TODO: support desktop-specific notification APIs if they provide a better
+	// experience. For example, the GNOME notification API.
+	notifySrv := notification.New(conn)
+
+	// TODO: this message needs to be crafted better as it's the only thing guaranteed to be delivered.
+	summary := fmt.Sprintf(i18n.G("Pending update of %q snap"), refreshInfo.InstanceName)
+	var urgencyLevel notification.Urgency
+	var body string
+	var icon string
+	var hints []notification.Hint
+	if daysLeft := int(refreshInfo.TimeRemaining.Truncate(time.Hour).Hours() / 24); daysLeft > 0 {
+		urgencyLevel = notification.LowUrgency
+		body = fmt.Sprintf(i18n.NG(
+			"Close the app to avoid disruptions (%d day left)",
+			"Close the app to avoid disruptions (%d days left)", daysLeft), daysLeft)
+	} else if hoursLeft := int(refreshInfo.TimeRemaining.Truncate(time.Minute).Minutes() / 60); hoursLeft > 0 {
+		urgencyLevel = notification.NormalUrgency
+		body = fmt.Sprintf(i18n.NG(
+			"Close the app to avoid disruptions (%d hour left)",
+			"Close the app to avoid disruptions (%d hours left)", hoursLeft), hoursLeft)
+	} else if minutesLeft := int(refreshInfo.TimeRemaining.Truncate(time.Minute).Minutes()); minutesLeft > 0 {
+		urgencyLevel = notification.CriticalUrgency
+		body = fmt.Sprintf(i18n.NG(
+			"Close the app to avoid disruptions (%d minute left)",
+			"Close the app to avoid disruptions (%d minutes left)", minutesLeft), minutesLeft)
+	} else {
+		summary = fmt.Sprintf("Snap %q is refreshing now!", refreshInfo.InstanceName)
+		urgencyLevel = notification.CriticalUrgency
+	}
+	hints = append(hints, notification.WithUrgency(urgencyLevel))
+	if refreshInfo.BusyAppDesktopEntry != "" {
+		hints = append(hints, notification.WithDesktopEntry(refreshInfo.BusyAppDesktopEntry))
+		// Extract the icon manually
+		parser := goconfigparser.New()
+		if err := parser.ReadFile(fmt.Sprintf("/var/lib/snapd/desktop/applications/%s.desktop", refreshInfo.BusyAppDesktopEntry)); err == nil {
+			icon, _ = parser.Get("Desktop Entry", "Icon")
+			fmt.Printf("found desktop file icon %q\n", icon)
+		}
+	}
+	msg := &notification.Message{
+		AppName:       refreshInfo.BusyAppName,
+		Summary:       summary,
+		Icon:          icon,
+		Body:          body,
+		ExpireTimeout: time.Second * 10,
```

Comment:
This value seems to be ignored in gnome shell. I need to do a pass around in kde and older unity to see what happens in practice.

Reply:
Removed